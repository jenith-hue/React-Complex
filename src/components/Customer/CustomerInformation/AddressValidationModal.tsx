import { RACModal, RACRadio } from '@rentacenter/racstrap';
import { RACButton, makeStyles } from '@rentacenter/racstrap';
import { Typography } from '@material-ui/core';
import React, { useState } from 'react';
import clsx from 'clsx';

export interface ModalProps {
  onClose: (onclose: boolean) => void;
  suggestedAddress?: Record<string, unknown>;
  enteredAddress?: Record<string, unknown>;
  handleSave: (onsave: string) => void;
  pending: boolean;
  text?: string;
}

export interface AddressNotValidatedModalProps {
  onClose: (onclose: boolean) => void;
  pending: boolean;
  text?: string;
}

export const useStyles = makeStyles((theme: any) => ({
  dialogContent: {
    marginBottom: '2rem',
  },
  dialogContent2: {
    textAlign: 'center',
    fontFamily: 'OpenSans-Bold',
  },
  dialogRoot: {
    '& .MuiPaper-rounded': {
      borderRadius: theme.typography.pxToRem(12),
    },
    '& .MuiDialog-paperWidthXs': {
      maxWidth: theme.typography.pxToRem(500),
    },
    '& .MuiTypography-h5': {
      fontFamily: 'OpenSans-semiBold',
      fontSize: theme.typography.pxToRem(20),
      fontWeight: 500,
      lineHeight: theme.typography.pxToRem(30),
    },
    '& .MuiTypography-h6': {
      fontFamily: 'OpenSans-semiBold',
      lineHeight: theme.typography.pxToRem(20),
    },
  },
  suggestedAddressesHeadline: {
    fontFamily: 'OpenSans-Bold !important',
    margin: '10px 0 !important',
  },
  suggestedAddressesInput: {
    fontSize: `${theme.typography.pxToRem(14)} !important`,
  },
  buttonRoot: {
    margin: 'auto',
  },
  cancelButton: {
    margin: 'auto',
    marginLeft: theme.typography.pxToRem(16),
  },
}));

export const AddressNotValidatedModal = ({
  onClose,
  text,
  pending,
}: AddressNotValidatedModalProps) => {
  const classes = useStyles();

  return (
    <RACModal
      isOpen
      classes={{
        dialogContent: classes.dialogContent2,
        dialog: classes.dialogRoot,
      }}
      maxWidth="sm"
      title="Validation Message"
      content={
        <>
          <Typography
            variant="body1"
            display="inline"
            className={classes.suggestedAddressesHeadline}
          >
            {text}
          </Typography>
          {!text?.includes('Zip') && (
            <Typography
              variant="body1"
              display="block"
              className={classes.suggestedAddressesHeadline}
            >
              Please Validate Address Manually
            </Typography>
          )}
        </>
      }
      onClose={() => onClose(false)}
      buttons={
        <>
          <RACButton
            variant="contained"
            color="primary"
            className={classes.buttonRoot}
            onClick={() => onClose(false)}
            disabled={pending}
            loading={pending}
          >
            Ok
          </RACButton>
        </>
      }
    />
  );
};

export const AddressValidationModal = ({
  onClose,
  suggestedAddress,
  enteredAddress,
  handleSave,
  pending,
}: ModalProps) => {
  const { StreetAddress, City, StateProvince, PostalCode } =
    suggestedAddress || {};

  const suggestedAddressString = `${StreetAddress}, ${City}, ${StateProvince} - ${PostalCode}`;

  const { addressLine1, addressLine2, city, state, postalCode } =
    enteredAddress || {};

  const addressLine2value = addressLine2 ? `${addressLine2}, ` : '';

  const enteredAddressString = `${addressLine1}, ${addressLine2value}${city}, ${state} - ${postalCode}`;

  const classes = useStyles();
  const [selectedSuggestedAddress, setSelectedSuggestedAddress] = useState(
    suggestedAddressString
  );
  const [selectedEnteredAddress, setSelectedEnteredAddress] = useState('');

  const handleSuggestedAddressSelect = (value: string) => {
    setSelectedSuggestedAddress(value);
    setSelectedEnteredAddress('');
  };

  const handleEnteredAddressSelect = (value: string) => {
    setSelectedEnteredAddress(value);
    setSelectedSuggestedAddress('');
  };

  const isOverrideDisabled = pending || selectedSuggestedAddress.length > 0;

  const isContinueDisabled = pending || selectedEnteredAddress.length > 0;

  const isOverrideLoading = pending && selectedEnteredAddress.length > 0;
  const isContinueLoading = pending && selectedSuggestedAddress.length > 0;

  return (
    <RACModal
      isOpen
      classes={{
        dialogContent: classes.dialogContent,
        dialog: classes.dialogRoot,
      }}
      maxWidth="xs"
      title="Validation Message"
      content={
        <>
          <Typography display="inline" variant="body2">
            Entered address is not located
          </Typography>
          <div>
            <Typography className={clsx(classes.suggestedAddressesHeadline)}>
              Suggested Addresses
            </Typography>
            <RACRadio
              name="SuggestedAddress"
              label={suggestedAddressString}
              value={suggestedAddressString}
              checked={
                selectedSuggestedAddress.length > 0 &&
                suggestedAddressString === selectedSuggestedAddress
              }
              onChange={(event) =>
                handleSuggestedAddressSelect(event.target.value)
              }
              classes={{
                label: classes.suggestedAddressesInput,
              }}
            />
          </div>
          <div>
            <Typography className={clsx(classes.suggestedAddressesHeadline)}>
              Entered Address
            </Typography>
            <RACRadio
              name="EnteredAddress"
              label={enteredAddressString}
              value={enteredAddressString}
              checked={
                selectedEnteredAddress.length > 0 &&
                enteredAddressString === selectedEnteredAddress
              }
              onChange={(event) =>
                handleEnteredAddressSelect(event.target.value)
              }
              classes={{
                label: classes.suggestedAddressesInput,
              }}
            />
          </div>
        </>
      }
      onClose={() => onClose(false)}
      buttons={
        <>
          <RACButton
            variant="outlined"
            color="secondary"
            onClick={() => onClose(false)}
            className={classes.cancelButton}
            disabled={pending}
          >
            Cancel
          </RACButton>
          <RACButton
            variant="contained"
            color="primary"
            onClick={() => handleSave('isEnteredAddress')}
            disabled={isOverrideDisabled}
            loading={isOverrideLoading}
          >
            Override
          </RACButton>
          <RACButton
            variant="contained"
            color="primary"
            onClick={() => handleSave('')}
            disabled={isContinueDisabled}
            loading={isContinueLoading}
          >
            Continue
          </RACButton>
        </>
      }
    />
  );
};
