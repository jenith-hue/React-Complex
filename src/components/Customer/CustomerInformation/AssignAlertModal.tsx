import {
  RACModal,
  RACCheckBox,
  RACButton,
  makeStyles,
  RACTextField,
} from '@rentacenter/racstrap';
import React, { useState, useEffect } from 'react';
import { AlertPayload } from '../../../types/types';
import { getCancelTokenSource } from '../../../api/client';
import { CancelTokenSource } from 'axios';
import {
  useCustomerAlerts,
  useCustomerAlertsActions,
} from '../../../context/CustomerAlert/CustomerAlertsProvider';
import { Option } from '../../../types/types';
import { REQUIRED_FIELD_MESSAGE } from '../../../constants/constants';

export interface ModalProps {
  onClose: (onclose: boolean) => void;
  onSave: () => void;
  open: boolean;
}

export const OTHER_ALERT_TYPE_ID = '3';

export const useStyles = makeStyles((theme: any) => ({
  dialogContent: {
    marginBottom: '2rem',
    padding: '0.5rem 1rem',
  },
  dialogRoot: {
    '& .MuiPaper-rounded': {
      borderRadius: theme.typography.pxToRem(12),
    },
    '& .MuiDialog-paperWidthXs': {
      maxWidth: theme.typography.pxToRem(500),
    },
    '& .MuiTypography-h5': {
      fontFamily: 'OpenSans-semiBold',
      fontSize: theme.typography.pxToRem(20),
      fontWeight: 500,
      lineHeight: theme.typography.pxToRem(30),
    },
    '& .MuiTypography-h6': {
      fontFamily: 'OpenSans-semiBold',
      lineHeight: theme.typography.pxToRem(20),
    },
    '& .MuiFormControlLabel-root': {
      display: 'flex',
    },
    '& .MuiFormControlLabel-label': {
      fontSize: theme.typography.pxToRem(16),
      fontFamily: 'OpenSans-regular',
    },
    '& .MuiOutlinedInput-multiline': {
      margin: '0.75rem !important',
    },
    '& p': {
      marginLeft: theme.typography.pxToRem(12),
    },
  },
  otherTextContent: {
    width: '100%',
  },
  flex: {
    display: 'flex',
    '& .MuiSvgIcon-root': { fontSize: 30 },
  },
}));

export const AssignAlertModal = ({ onClose, onSave }: ModalProps) => {
  const classes = useStyles();

  const [assignAlertModalValues, setAssignAlertModalValues] = useState<
    Option[]
  >([]);
  const { customerAlerts, loading, allAlerts } = useCustomerAlerts();
  const { saveCustomerAlerts } = useCustomerAlertsActions();
  const getSavedAlerts = () => {
    if (customerAlerts?.length) {
      return (
        customerAlerts
          .filter(
            (alert) =>
              alert.alertCleared === 0 &&
              alert.alertTypeId !== OTHER_ALERT_TYPE_ID
          )
          ?.map((alert) => String(alert.alertTypeId)) || []
      );
    }
    return [];
  };

  const [selectedAlerts, setSelectedAlerts] = useState<string[]>(
    getSavedAlerts()
  );

  const [otherTextContent, setOtherTextContent] = useState('');

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const alertName = event.target.name;
    if (!selectedAlerts.includes(alertName)) {
      const selected = [...selectedAlerts, alertName];
      setSelectedAlerts(selected);
    } else {
      const selected = selectedAlerts.filter((alert) => alert !== alertName);
      setSelectedAlerts(selected);
    }
  };

  useEffect(() => {
    setAssignAlertModalValues(
      allAlerts.map(
        (alert) =>
          ({
            value: alert.alertTypeId,
            label: alert.alertTypeDescEn,
          } as Option)
      )
    );
  }, [allAlerts]);
  useEffect(() => {
    const alerts = getSavedAlerts();
    alerts?.length && setSelectedAlerts(alerts);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [customerAlerts]);

  const buildAlertTypesPayload = (): AlertPayload[] => {
    const savedAlertIds = customerAlerts
      .filter((alert) => alert.alertTypeId !== OTHER_ALERT_TYPE_ID)
      ?.map((alert) => alert.alertTypeId);

    return selectedAlerts?.map((alert) => {
      let alertText = '';
      if (alert === OTHER_ALERT_TYPE_ID) {
        // If option is other
        alertText = otherTextContent;
      } else {
        alertText =
          assignAlertModalValues.find((option) => option.value === alert)
            ?.label || '';
      }

      if (savedAlertIds.includes(alert)) {
        const savedAlert = customerAlerts.find(
          (savedCustomerAlert) => savedCustomerAlert.alertTypeId === alert
        );

        if (savedAlert) {
          return {
            alertTypeId: Number(alert),
            customerAlertId: Number(savedAlert?.customerAlertId),
            alertClear: 0,
            alertText: alertText,
          };
        }
      }

      return {
        alertTypeId: Number(alert),
        customerAlertId: null,
        alertClear: 0,
        alertText: alertText,
      };
    });
  };

  const handleSaveAlert = async () => {
    if (
      selectedAlerts.includes(OTHER_ALERT_TYPE_ID) &&
      otherTextContent.trim() === ''
    )
      return;
    const cancelToken: CancelTokenSource = getCancelTokenSource();
    const customerAlertPayload = buildAlertTypesPayload();
    saveCustomerAlerts(customerAlertPayload, cancelToken).finally(onSave);
  };

  const alertAlreadyExist = (alertId: string) =>
    alertId !== OTHER_ALERT_TYPE_ID &&
    customerAlerts?.find(
      (alert) => alert.alertTypeId === alertId && alert.alertCleared === 0
    );

  return (
    <RACModal
      isOpen
      classes={{
        dialogContent: classes.dialogContent,
        dialog: classes.dialogRoot,
      }}
      maxWidth="xs"
      title="Customer Alerts"
      content={
        <>
          {assignAlertModalValues?.map((alerts, index) => {
            return (
              <RACCheckBox
                label={alerts.label}
                size="medium"
                checked={selectedAlerts.includes(alerts.value)}
                disabled={!!alertAlreadyExist(alerts.value)}
                onChange={handleChange}
                name={alerts.value}
                key={index}
                classes={{ checkbox: classes.flex }}
              />
            );
          })}

          {selectedAlerts.includes(OTHER_ALERT_TYPE_ID) && (
            <RACTextField
              classes={{ input: classes.otherTextContent }}
              // eslint-disable-next-line @typescript-eslint/ban-ts-comment
              //@ts-ignore
              multiline
              variant="outlined"
              maxRows={3}
              minRows={3}
              value={otherTextContent}
              onChange={(value) => {
                setOtherTextContent(value);
              }}
              errorMessage={
                otherTextContent.trim() === '' ? REQUIRED_FIELD_MESSAGE : ''
              }
            />
          )}
        </>
      }
      onClose={() => onClose(false)}
      buttons={
        <>
          <RACButton
            variant="outlined"
            color="secondary"
            onClick={() => onClose(false)}
            disabled={loading}
          >
            Cancel
          </RACButton>
          <RACButton
            variant="contained"
            color="primary"
            disabled={loading}
            loading={loading}
            onClick={handleSaveAlert}
          >
            Save
          </RACButton>
        </>
      }
    />
  );
};
