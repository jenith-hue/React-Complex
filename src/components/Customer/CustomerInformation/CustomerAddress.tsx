import {
  Grid,
  makeStyles,
  Typography,
  RACTextField,
  RACButton,
  RACSelect,
} from '@rentacenter/racstrap';
import React, { useEffect, useState } from 'react';
import { orderBy } from 'lodash';
import { useCustomerDetails } from '../../../context/CustomerDetails/CustomerDetailsProvider';
import { useStoreDetails } from '../../../context/Store/StoreProvider';
import { AddressType } from '../../../types/types';
import {
  AddressNotValidatedModal,
  AddressValidationModal,
} from '../../Customer/CustomerInformation/AddressValidationModal';
import { format } from 'date-fns';
import * as api from '../../../api/Customer';
import {
  getSelectedStore,
  randomNumericStringGenerator,
} from '../../../utils/utils';
import { API_ERROR_MESSAGE } from '../../../constants/constants';
import { ApiErrorModal } from '../../../common/ApiError/ApiError';

export const customerAddressTestId = 'customerAddressTestId';
export const customerAddressLine1TestId = 'customerAddressLine1TestId';
export const customerAddressLine2TestId = 'customerAddressLine2TestId';
export const customerPostalCodeTestId = 'customerPostalCodeTestId';
export const customerCityTestId = 'customerCityTestId';
export const customerStateTestId = 'customerStateTestId';
export const customerAddressFormId = 'customerAddressFormId';

enum ErrorMessage {
  INVALID_FORMAT_ERROR_MESSAGE = 'Postal code should be either in *****-**** or ***** format!',
  MAX_LENGTH_ERROR_MESSAGE = 'Max length exceeded!',
  ALL_0_ERROR_MESSAGE = `Postal code value can't be all 0's!`,
}

const useClasses = makeStyles((theme: any) => ({
  customerAddress: {
    display: 'flex',
    flexDirection: 'column',
    padding: `${theme.typography.pxToRem(0)} ${theme.typography.pxToRem(12)}`,
  },
  addressLineWrapper: { display: 'flex', flexDirection: 'row' },
  saveButton: {
    height: theme.typography.pxToRem(43),
    marginTop: theme.typography.pxToRem(16),
  },
  row: { marginBottom: theme.typography.pxToRem(24) },
  selectOptionsPaper: {
    maxHeight: theme.typography.pxToRem(200),
  },
}));

enum FormFields {
  city = 'city',
  state = 'state',
  postalCode = 'postalCode',
  addressLine1 = 'addressLine1',
  addressLine2 = 'addressLine2',
  addressId = 'addressId',
}

export const CustomerAddress = () => {
  const classes = useClasses();

  const {
    customerDetails: { addresses, customerId },
  } = useCustomerDetails();

  const [
    openCustomerAddressValidationModal,
    setOpenCustomerAddressValidationModal,
  ] = useState<boolean>(false);

  const [
    openCustomerAddressNotValidatedModal,
    setOpenCustomerAddressNotValidatedModal,
  ] = useState<boolean>(false);

  const [suggestedAddress, setSuggestedAddress] = useState<any>({});
  const [enteredAddress, setEnteredAddress] = useState<any>({});

  const [pending, setPending] = useState<boolean>(false);
  const [hasApiErrorUpdateCustomer, setHasApiErrorUpdateCustomer] =
    useState<boolean>(false);

  const primaryAddress = addresses?.filter(
    (address) => address.addressType === AddressType.PRIMARY
  );

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const customerAddress =
    primaryAddress && primaryAddress.length > 0
      ? primaryAddress[0]
      : {
          city: '',
          state: '',
          postalCode: '',
          addressLine1: '',
          addressLine2: '',
          addressId: '',
        };

  const [customerAddressLine1, setCustomerAddressLine1] = useState('');
  const [customerAddressLine2, setCustomerAddressLine2] = useState('');
  const [stateProvince, setStateProvince] = useState('');
  const [customerPostalCode, setCustomerPostalCode] = useState('');
  const [customerCity, setCustomerCity] = useState('');

  useEffect(() => {
    const { addressLine1, addressLine2, city, state, postalCode } =
      customerAddress;

    setCustomerAddressLine1(addressLine1);
    setCustomerAddressLine2(addressLine2);
    setCustomerCity(city);
    setStateProvince(state);
    setCustomerPostalCode(postalCode);
  }, [customerAddress]);

  const { store, isAddressDoctorEnabled } = useStoreDetails();

  const countryAbbreviation = store?.countryAbb;

  const storeNumber = getSelectedStore();

  const [usStatesList, setUSStatesList] = useState([]);

  const [addressText, setAddressText] = useState('');

  const [postalCodeErrorMessage, setPostalCodeErrorMessage] = useState('');
  const [hasPostalCodeError, setHasPostalCodeError] = useState(false);
  const [hasStatesApiError, setHasStatesApiError] = useState(false);
  const [fetchingData, setFetchingData] = useState(false);

  useEffect(() => {
    setFetchingData(true);
    api
      .getCustomerStatesList()
      .then((response) => {
        const statesList = response.map((state: any) => {
          return { label: state.abbreviation, value: state.abbreviation };
        });
        const sortedStatesList: any = orderBy(statesList, ['label'], ['asc']);

        setUSStatesList(sortedStatesList);
        setHasStatesApiError(false);
      })
      .catch(() => {
        setHasStatesApiError(true);
      })
      .finally(() => {
        setFetchingData(false);
      });
  }, []);

  const handleSubmit = () => {
    const { addressId } = customerAddress;
    const currentTimeZoneOffset = format(new Date(), 'XXX');
    const currentTimeStamp = format(new Date(), 'yyyy-MM-dd-HH:mm:ss:SSSSSS');

    const getRegionID = () => {
      if (countryAbbreviation === 'US') {
        return '1';
      } else if (countryAbbreviation === 'MX') {
        return '2';
      }
      return '';
    };

    const enteredAddress = {
      addressLine1: customerAddressLine1,
      addressLine2: customerAddressLine2,
      state: stateProvince,
      postalCode: customerPostalCode,
      city: customerCity,
    };

    setEnteredAddress(enteredAddress);

    if (!isAddressDoctorEnabled) {
      handleSave('isEnteredAddress', enteredAddress);
      return;
    }

    setPending(true);

    const payload = {
      MessageID: `CMS006-${randomNumericStringGenerator()}`,
      MessageType: 'CMS006',
      MessageDTS: currentTimeStamp,
      MessageDTSTZOffset: currentTimeZoneOffset,
      RegionID: getRegionID(),
      ClientID: '8',
      ClientLocationNumber: storeNumber,
      ClientSource: '1',
      ClientOriginator: '1',
      LocationNumber: storeNumber,
      EncryptionMethod: '0',
      Addresses: [
        {
          AddressID: addressId,
          StreetAddress:
            customerAddressLine2?.trim().length > 0
              ? `${customerAddressLine1}, ${customerAddressLine2}`
              : customerAddressLine1,
          City: customerCity,
          StateProvince: stateProvince,
          PostalCode: customerPostalCode.replace('-', ''),
        },
      ],
      StoreConfig: {
        storeNumbers: [storeNumber],
        paramKeyNames: ['AddressDoctorEnabled'],
      },
    };

    api
      .addressValidator(payload)
      .then((response) => {
        const addresses = response?.validateAddress?.Addresses;
        if (addresses.MatchCode === '1') {
          setOpenCustomerAddressNotValidatedModal(true);
          setSuggestedAddress({});
          setAddressText(
            `Zip code ${customerPostalCode} doesn't exist in the state ${stateProvince}.`
          );
        } else if (
          addresses.MatchCode === '6' ||
          addresses.MatchCode === '5' ||
          addresses.MatchCode === '8'
        ) {
          setSuggestedAddress(addresses);
          setOpenCustomerAddressValidationModal(true);
        } else if (addresses.MatchCode === '2') {
          setOpenCustomerAddressNotValidatedModal(true);
          setAddressText(
            `Zipcode found: ${customerPostalCode}. Do you want to override?`
          );
        } else if (addresses.MatchCode === '10') {
          handleSave('isEnteredAddress', enteredAddress);
        }
      })
      .catch(() => {
        setAddressText(
          `Automatic Address Validation is Unavailable for store ${storeNumber}`
        );
        setSuggestedAddress({});
        setOpenCustomerAddressNotValidatedModal(true);
      })
      .finally(() => {
        setPending(false);
      });
  };

  const handleSave = (
    isEnteredAddressSelected: string,
    userInputAddress?: Record<string, unknown>
  ) => {
    setPending(true);
    let { addressID, addressLine1, addressLine2, city, state, postalCode } =
      userInputAddress || enteredAddress || {};

    const { AddressID, StreetAddress, City, StateProvince, PostalCode } =
      suggestedAddress || {};
    const isEnteredAddress = isEnteredAddressSelected.length > 0;
    if (!isEnteredAddress) {
      addressID = AddressID;
      addressLine1 = StreetAddress;
      addressLine2 = '';
      city = City;
      state = StateProvince;
      postalCode = PostalCode;
    }
    const payload = {
      customerId: customerId,
      addresses: [
        {
          addressType: 'PRIM',
          addressLine1,
          addressLine2,
          city,
          state,
          postalCode,
          addressId: addressID,
        },
      ],
    };
    api
      .updateCustomer(payload)
      .then(() => {
        setCustomerAddressLine1(addressLine1);
        setCustomerAddressLine2(addressLine2);
        setCustomerCity(city);
        setStateProvince(state);
        setCustomerPostalCode(postalCode);
      })
      .catch(() => {
        setHasApiErrorUpdateCustomer(true);
      })
      .finally(() => {
        setPending(false);
        setOpenCustomerAddressNotValidatedModal(false);
        setOpenCustomerAddressValidationModal(false);
      });
  };

  const handleZipChange = (value: string) => {
    const regExp = /(^\d{5}$)|(^\d{5}-\d{4}$)/;
    setCustomerPostalCode(value);
    if (value?.length > 10) {
      setHasPostalCodeError(true);
      setPostalCodeErrorMessage(ErrorMessage.MAX_LENGTH_ERROR_MESSAGE);
    } else if (value.includes('00000') || value.includes('-0000')) {
      setHasPostalCodeError(true);
      setPostalCodeErrorMessage(ErrorMessage.ALL_0_ERROR_MESSAGE);
    } else if (!regExp.test(value)) {
      setHasPostalCodeError(true);
      setPostalCodeErrorMessage(ErrorMessage.INVALID_FORMAT_ERROR_MESSAGE);
    } else {
      setHasPostalCodeError(false);
    }
  };

  const isDirty = () => {
    const { addressLine1, addressLine2, city, state, postalCode } =
      customerAddress;
    if (
      addressLine1 === customerAddressLine1 &&
      addressLine2 === customerAddressLine2 &&
      city === customerCity &&
      state === stateProvince &&
      postalCode === customerPostalCode
    ) {
      return false;
    }
    return true;
  };

  const isFormValid = () => {
    if (
      customerAddressLine1.trim().length > 0 &&
      customerCity.trim().length > 0
    ) {
      return true;
    }
    return false;
  };

  return (
    <div
      className={classes.customerAddress}
      data-testid={customerAddressTestId}
    >
      <Grid container>
        <Grid item sm={12} md={12} lg={12} classes={{ root: classes.row }}>
          <Typography variant="h4">Address</Typography>
        </Grid>
        <Grid item sm={12} md={12} lg={12}>
          <Grid container spacing={2} classes={{ root: classes.row }}>
            <Grid item sm={6} md={6} lg={6}>
              <RACTextField
                name={FormFields.addressLine1}
                label="Address Line 1"
                value={customerAddressLine1}
                type="text"
                required
                onChange={(value) => setCustomerAddressLine1(value)}
              />
            </Grid>
            <Grid item sm={6} md={6} lg={6}>
              <RACTextField
                name={FormFields.addressLine2}
                label="Address Line 2"
                value={customerAddressLine2}
                type="text"
                onChange={(value) => setCustomerAddressLine2(value)}
              />
            </Grid>
          </Grid>
          <Grid container spacing={2} classes={{ root: classes.row }}>
            <Grid item sm={10} md={10} lg={10}>
              <Grid container spacing={2}>
                <Grid item sm={4} md={4} lg={4}>
                  <RACTextField
                    name={FormFields.postalCode}
                    label="Zip"
                    value={customerPostalCode}
                    type="text"
                    required
                    onChange={(value) => handleZipChange(value)}
                    {...(hasPostalCodeError && {
                      errorMessage: postalCodeErrorMessage,
                    })}
                  />
                </Grid>
                <Grid item sm={4} md={4} lg={4}>
                  <RACTextField
                    name={FormFields.city}
                    label="City"
                    value={customerCity}
                    type="text"
                    required
                    onChange={(city) => setCustomerCity(city)}
                  />
                </Grid>
                <Grid item sm={4} md={4} lg={4}>
                  <RACSelect
                    classes={{ paper: classes.selectOptionsPaper }}
                    name={FormFields.state}
                    inputLabel="State"
                    defaultValue={stateProvince}
                    options={usStatesList}
                    required
                    onChange={(e) => {
                      setStateProvince(e.target.value);
                    }}
                    loading={fetchingData}
                    {...(hasStatesApiError && {
                      errorMessage: API_ERROR_MESSAGE,
                    })}
                  />
                </Grid>
              </Grid>
            </Grid>
            <Grid item sm={2} md={2} lg={2} container justifyContent="flex-end">
              <RACButton
                variant="contained"
                disabled={
                  !isFormValid() ||
                  !isDirty() ||
                  hasPostalCodeError ||
                  pending ||
                  hasStatesApiError
                }
                size="large"
                color="primary"
                loading={pending}
                className={classes.saveButton}
                onClick={() => handleSubmit()}
              >
                Save
              </RACButton>
            </Grid>
            {hasApiErrorUpdateCustomer && (
              <ApiErrorModal
                open
                message="Failed to update address!"
                onClose={() => setHasApiErrorUpdateCustomer(false)}
              />
            )}
            {openCustomerAddressValidationModal && (
              <AddressValidationModal
                handleSave={(onsave) => handleSave(onsave)}
                suggestedAddress={suggestedAddress}
                enteredAddress={enteredAddress}
                pending={pending}
                onClose={() => setOpenCustomerAddressValidationModal(false)}
              />
            )}
            {openCustomerAddressNotValidatedModal && (
              <AddressNotValidatedModal
                pending={pending}
                text={addressText}
                onClose={() => setOpenCustomerAddressNotValidatedModal(false)}
              />
            )}
          </Grid>
        </Grid>
      </Grid>
    </div>
  );
};
