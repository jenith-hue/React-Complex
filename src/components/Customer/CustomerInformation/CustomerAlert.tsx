import React from 'react';
import { RACChip, makeStyles } from '@rentacenter/racstrap';
import {
  useCustomerAlerts,
  useCustomerAlertsActions,
} from '../../../context/CustomerAlert/CustomerAlertsProvider';

export const customerAlertTestId = 'customerAlertTestId';

const useClasses = makeStyles((theme: any) => ({
  customerAlert: {
    display: 'flex',
    flexGrow: 1,
    backgroundColor: 'white',
    overflowX: 'auto',
    padding: theme.typography.pxToRem(8),
    marginRight: theme.typography.pxToRem(10),
  },
  chip: {
    marginRight: '.5rem',
  },
}));

export const CustomerAlert_ = () => {
  const classes = useClasses();
  const { customerAlerts } = useCustomerAlerts();
  const { removeCustomerAlert } = useCustomerAlertsActions();

  return (
    <div className={classes.customerAlert} data-testid={customerAlertTestId}>
      {customerAlerts
        ?.filter((alert) => alert.alertCleared === 0)
        ?.map((alert) => (
          <RACChip
            classes={{ root: classes.chip }}
            key={alert.customerAlertId}
            label={alert.alertTypeDescEn}
            onDelete={() => removeCustomerAlert(alert)}
          />
        ))}
    </div>
  );
};
