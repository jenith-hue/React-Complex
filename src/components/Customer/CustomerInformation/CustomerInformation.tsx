import React from 'react';
import { RACAccordion, makeStyles } from '@rentacenter/racstrap';
import { CustomerInformationDetails } from './CustomerInformationDetails';

const useStyles = makeStyles(() => ({
  customerInformationRoot: {
    width: 'calc(100% + 1.5rem)',
    left: '-0.75rem',
    margin: '0 !important',
  },
  accordionSummaryRoot: {
    width: '10px',
    position: 'absolute',
    right: 10,
    top: '-70px',
  },
  accordionSummaryContent: {
    margin: '0 !important',
  },
}));

export const CustomerInformation = () => {
  const classes = useStyles();

  return (
    <>
      <RACAccordion
        details={<CustomerInformationDetails />}
        accordionProps={{
          classes: {
            root: classes.customerInformationRoot,
          },
        }}
        summaryProps={{
          classes: {
            root: classes.accordionSummaryRoot,
            content: classes.accordionSummaryContent,
          },
        }}
      />
    </>
  );
};
