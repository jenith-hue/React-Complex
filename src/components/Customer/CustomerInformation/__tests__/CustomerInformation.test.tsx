/* eslint-disable react/display-name */

import React from 'react';
import renderer from 'react-test-renderer';
import { CustomerInformation } from '../CustomerInformation';

interface PropWithChildren {
  children: JSX.Element;
}
jest.mock('@rentacenter/racstrap', () => ({
  RACAccordion: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACAccordion" {...rest}>
      {children}
    </div>
  ),
  makeStyles: () => () => ({}),
}));

jest.mock('../CustomerInformationDetails', () => ({
  CustomerInformationDetails: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="CustomerInformationDetails" {...rest}>
      {children}
    </div>
  ),
  makeStyles: () => () => ({}),
}));

describe('CustomerInformation', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<CustomerInformation />);

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
