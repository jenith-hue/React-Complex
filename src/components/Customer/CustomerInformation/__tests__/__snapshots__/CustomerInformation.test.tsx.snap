// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerInformation renders correctly 1`] = `
<div
  accordionProps={
    Object {
      "classes": Object {
        "root": undefined,
      },
    }
  }
  data-testid="RACAccordion"
  details={<CustomerInformationDetails />}
  summaryProps={
    Object {
      "classes": Object {
        "content": undefined,
        "root": undefined,
      },
    }
  }
/>
`;
